extensions/ezmsg-sigproc/src/ezmsg/sigproc/affinetransform.py

```diff
@@ -48,6 +51,43 @@ def affine_transform(
         else:
             axis_idx = axis_arr_in.get_axis_idx(axis)
 
+        if b_first:
+            # First data sample. Determine if we need to modify the transformed axis.
+            b_first = False
+            if (
+                    axis in axis_arr_in.axes
+                    and hasattr(axis_arr_in.axes[axis], "labels")
+                    and weights.shape[0] != weights.shape[1]
+            ):
+                in_labels = axis_arr_in.axes[axis].labels
+                new_labels = []
+                n_in = weights.shape[1 if right_multiply else 0]
+                n_out = weights.shape[0 if right_multiply else 1]
+                if len(in_labels) != n_in:
+                    # Something upstream did something it wasn't supposed to. We will drop the labels.
+                    ez.logger.warning(f"Received {len(in_labels)} for {n_in} inputs. Check upstream labels.")
+                else:
+                    b_used_inputs = np.any(weights, axis=0 if right_multiply else 1)
+                    b_filled_outputs = np.any(weights, axis=1 if right_multiply else 0)
+                    if np.all(b_used_inputs) and np.all(b_filled_outputs):
+                        # All inputs are used and all outputs are used, but n_in != n_out.
+                        # Mapping cannot be determined.
+                        new_labels = []
+                    elif np.all(b_used_inputs):
+                        # Strange scenario: New outputs are filled with empty data.
+                        in_ix = 0
+                        new_labels = []
+                        for out_ix in range(n_out):
+                            if b_filled_outputs[out_ix]:
+                                new_labels.append(in_labels[in_ix])
+                                in_ix += 1
+                            else:
+                                new_labels.append("")
```

Comment:
Maybe append `None` here instead? If you keep `""`, we should be explicit that we assume labels will always be a `typing.List[str]` if it exists

Reply:
Yes, it is always a `typing.List[str]`, not a `typing.List[typing.Optional[str]]`. The `""` is intended to indicate that we don't know what the label is. If anything, I was thinking about replacing the `""` with `"?"` or even `str(len(new_labels)+1)`. But I decided against these because integers might already have a different meaning, and I just know that "?" is going to interfere with somebody's localization or other parser down the line eventually.